### PR TITLE
Move StorageClass to a storage group

### PR DESCRIPTION
**Pull request by deads2k**
We discussed the pros and cons in sig-api-machinery yesterday.  Choosing a particular group name means that clients (including our internal code) require less work and re-swizzling to handle promotions between versions.  Even if you choose a group you end up not liking, the amount of work remains the same as the incubator work case: you move the affected kind, resource, and storage.

This moves the `StorageClass` type to the `storage.k8s.io` group (named for consistency with authentication, authorization, rbac, and imagepolicy).  There are two commits, one for manaul changes and one for generated code.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31886)

### CODE REVIEW

**Comment by deads2k** [comment]
@k8s-bot test this: issue #28285

**Comment by thockin** [comment]
Given the state of the 1.4 release can we really justify this move?  I am fine with it FOR THIS CASE, though I really do want to get some consensus on the more complex case of adding a beta type in a non-beta group, vs a catch-all incubator group.

@pwittrock 

@bgrant0607 FYI

**Comment by deads2k** [comment]
@lavalamp @thockin @pmorie @kubernetes/sig-storage I've gotten all the bits re-swizzled to move `StorageClass` to a storage group.  I'd really like to get this in 1.4 so that we can do a straight move instead of a migration.  ptal.

**Comment by wattsteve** [comment]
To the best of my knowledge, this change is completely blind-siding @kubernetes/sig-storage. I would have liked to have seen (or if there was one, been made aware of) a proposal for this change.  I'd like the storage SIG to have an opportunity to discuss/review this before there are any plans of merging this  PR

**Comment by deads2k** [comment]
> To the best of my knowledge, this change is completely blind-siding @kubernetes/sig-storage. I would have liked to have seen (or if there was one, been made aware of) a proposal for this change.

Sorry, it was discussed here https://github.com/kubernetes/kubernetes/pull/29694#issuecomment-242748918 (original pull) and here https://github.com/kubernetes/kubernetes/issues/31521 (issue opened) with tags for the original author and the active reviewers on that pull for the types.

**Comment by thockin** [comment]
Steve,  That's reasonable, though I don't expect any concern - it's purely
about naming.

**Comment by rootfs** [comment]
@deads2k you have to update the examples 
https://github.com/kubernetes/kubernetes/tree/master/examples/experimental/persistent-volume-provisioning

**Comment by wattsteve** [comment]
I think naming is important. These terms we choose carry a lot of context with them. It takes time for these terms and the context behind them to diffuse within our community and users. I'm sensitive about Storage Classes specifically because its not particularly well understood yet.

The Storage SIG has started thinking about naming (we discussed this at the Storage F2F) and one of the outcomes has been https://github.com/kubernetes/kubernetes/issues/30474. I think @jingxu97 hasn't seen much activity on that yet, because its lower priority than closing out Kube 1.4

**Comment by wattsteve** [comment]
OK folks, I'm reverting my objections. For reference, I misinterpreted this PR as renaming the StorageClass API Object, which is not what is being proposed.

**Comment by smarterclayton** [comment]
So the release justification for 1.4 is that this is a net new resource (no broken backwards users) and this is a signature 1.4 feature that we want to ensure reduces significant long term maintenance costs on.  So even though this increases 1.4 risk, it is our last chance to avoid significant future work.

**Comment by smarterclayton** [comment]
@pwittrock FYI for release process and exception process.  I don't have the handy list of the exception questions so I tried to handle them above.

**Comment by childsb** [comment]
+1 to this refactor for 1.4

**Comment by pwittrock** [comment]
@smarterclayton Thanks for messaging me on this.  This is coming in really late for 1.4.  I understand why you want to get it in.  What are the risks of accepting this?  If we do accept this for 1.4, I will ask that beforehand 2 people perform a thorough manual QA on the affected pieces to make sure these changes will not break anything. 

@derekwaynecarr do you think this is something that qualifies for the exception audit process?

**Comment by pmorie** [comment]
I personally think this qualifies as an exception to the audit process and will be happy to answer The Questions.

**Comment by pmorie** [comment]
The Three Questions -- my take.

1) What is the worst case scenario in accepting the change - what are the risks?

The risk here is limited to functional defects in the new storage API group and StorageClass API resource. The risk is low because we have not released this API resource before.

2) How difficult is it to undo the change if we discover an issue before the release

The manual change commit can be reverted and generated code regenerated.

3) What critical value does this give that is needed for v1.4?

Accepting this changes means that we will be releasing this new API resource in the right API group.  It will ease things for clients down the road because we believe we will eventually need to put this API resource into its own group to effectively create ACLs and other policy-based features around it.

**Comment by derekwaynecarr** [comment]
@pwittrock - this feels more like a bug than a feature.  getting naming right to me is different than supporting a new concept altogether.

**Comment by pwittrock** [comment]
@derekwaynecarr Thanks.  I tend to agree, but wanted to make sure I was being consistent with what we discussed this morning.

**Comment by pwittrock** [comment]
I think we should get this in, but ask that it is manually tested as part of the process.

**Comment by pwittrock** [comment]
Let me know if you need anything else from me on this.

File: pkg/master/storage_storage.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"k8s.io/kubernetes/pkg/api/rest"
+	storageapi "k8s.io/kubernetes/pkg/apis/storage"
+	storageapiv1beta1 "k8s.io/kubernetes/pkg/apis/storage/v1beta1"
+	"k8s.io/kubernetes/pkg/genericapiserver"
+	storageclassetcd "k8s.io/kubernetes/pkg/registry/storageclass/etcd"
+)
+
+type StorageRESTStorageProvider struct {
```
Comment (lavalamp) [inline pkg/master/storage_storage.go]: sigh...

**Comment by lavalamp** [comment]
I reviewed this and did not see any problems, however if something was missed I probably wouldn't have noticed. I think @krousey is out today, @caesarxuchao do you have time to make a review pass, too?

File: pkg/master/storage_storage.go (hunk 1 above)
Comment (deads2k) [inline pkg/master/storage_storage.go]: > sigh...

I try to imagine a beautiful new world with APIGroups having distinct packages for their `RESTStorage` and it helps me get through names like this.

**Comment by caesarxuchao** [comment]
I'm taking a look now. Sorry for the delay.

**Comment by caesarxuchao** [comment]
LGTM. Thanks @deads2k !

**Comment by deads2k** [comment]
@pmorie can you walk me through/send me a link to any special process I need to follow for 1.4 picks?

**Comment by pmorie** [comment]
@deads2k I'll walk you through in meatspace if you need it, cherry-pick doc is at: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/cherry-picks.md

**Comment by lavalamp** [comment]
You don't need anything special if this merges before we do a fast-forward.

**Comment by wojtek-t** [comment]
This PR broke gke-slow suite, concretely this test:
https://github.com/kubernetes/kubernetes/issues/32185

@k8s-oncall - FYI

**Comment by pwittrock** [comment]
@smarterclayton @lavalamp Can you make sure the test break gets resolved.  I do not plan to cherrypick this until then.

**Comment by eparis** [comment]
The test break was 'fixed' in https://github.com/kubernetes/kubernetes/pull/32199
So these can/should be picked together. Once GKE is fixed we can decide if we want to pick https://github.com/kubernetes/kubernetes/pull/32200

**Comment by pwittrock** [comment]
@eparis Thanks

**Comment by jlowdermilk** [comment]
Actually, @pwittrock, there's a chicken and egg problem here. Until this api group exists in 1.4, we can't cleanly turn it on in GKE. This change needs to be cherrypicked first.

**Comment by pwittrock** [comment]
@jlowdermilk This will be cherrypicked into 1.4 shortly.

**Pull request merged**